index.js

```diff
@@ -155,179 +145,142 @@ function readProtocolData (reader) {
   return protocolData
 }
 
-function serializeAck (requestId, protocolData) {
-  const writer = new Writer()
-  writeProtocolData(writer, protocolData)
+function writeError (writer, data) {
+  const ilpPacket = maybeSerializeIlpError(data.rejectionReason)
+  writer.write(ilpPacket)
+  writeProtocolData(writer, data.protocolData)
+}
 
-  return writeEnvelope(TYPE_ACK, requestId, writer.getBuffer())
+function writePrepare (writer, data) {
+  const transferIdBuffer = Buffer.from(data.transferId.replace(/-/g, ''), 'hex')
+  const amountAsPair = stringToTwoNumbers(data.amount)
+  const executionConditionBuffer = Buffer.from(data.executionCondition, 'base64')
+  const expiresAtBuffer = toGeneralizedTimeBuffer(data.expiresAt)
+  writer.write(transferIdBuffer)
+  writer.writeUInt64(amountAsPair)
+  writer.write(executionConditionBuffer)
+  writer.writeVarOctetString(expiresAtBuffer)
+  writeProtocolData(writer, data.protocolData)
 }
 
-function deserializeAck (buffer) {
-  const { requestId, data } = readEnvelope(buffer)
-  const reader = new Reader(data)
+function writeFulfill (writer, data) {
+  const transferIdBuffer = Buffer.from(data.transferId.replace(/-/g, ''), 'hex')
+  const fulfillmentBuffer = Buffer.from(data.fulfillment, 'base64')
+  writer.write(transferIdBuffer)
+  writer.write(fulfillmentBuffer)
+  writeProtocolData(writer, data.protocolData)
+}
 
-  const protocolData = readProtocolData(reader)
-  return { requestId, protocolData }
+function writeReject (writer, data) {
+  const transferIdBuffer = Buffer.from(data.transferId.replace(/-/g, ''), 'hex')
+  const rejectionReasonBuffer = maybeSerializeIlpError(data.rejectionReason)
+  writer.write(transferIdBuffer)
+  writer.write(rejectionReasonBuffer)
+  writeProtocolData(writer, data.protocolData)
 }
 
-function serializeResponse (requestId, protocolData) {
+function serialize (obj) {
   const writer = new Writer()
```

Comment:
maybe this should be called `contentsWriter`; having writer and envelopewriter is a bit confusing